File: Account.cpp
#include "Account.h"
#include "Date.h"

//디폴트 생성자
Account::Account()
	:briefs(""), remarks("")
{
	this->amount = 0;
	this->balance = 0;
}

//매개변수 5개 생성자
Account::Account(Date date, string briefs, Currency amount, Currency balance, string remarks)
	:date(date), briefs(briefs), remarks(remarks)
{
	this->amount = amount;
	this->balance = balance;
}

//복사생성자
Account::Account(const Account& source)
	:date(source.date), briefs(source.briefs), remarks(source.remarks)
{
	this->amount = source.amount;
	this->balance = source.balance;
}

//치환연산자
Account& Account::operator=(const Account& source)
{
	this->date = source.date;
	this->briefs = source.briefs;
	this->amount = source.amount;
	this->balance = source.balance;
	this->remarks = source.remarks;

	return *this;
}

//소멸자
Account::~Account()
{

}

//IsEqual
bool Account::IsEqual(const Account& other)
{
	bool ret = false;
	if (this->date.IsEqual(other.date) == true && this->briefs.compare(other.briefs) == 0 &&
		this->amount == other.amount && this->balance == other.balance &&
		this->remarks.compare(other.remarks) == 0)
	{
		ret = true;
	}
	return ret;
}

//IsNotEqual
bool Account::IsNotEqual(const Account& other)
{
	bool ret = false;
	if (this->date.IsEqual(other.date) != true || this->briefs.compare(other.briefs) != 0 ||
		this->amount != other.amount || this->balance != other.balance ||
		this->remarks.compare(other.remarks) != 0)
	{
		ret = true;
	}
	return ret;
}

//operator==
bool Account::operator==(const Account& other)
{
	bool ret = false;
	if (this->date == other.date && this->briefs.compare(other.briefs) == 0 &&
		this->amount == other.amount && this->balance == other.balance &&
		this->remarks.compare(other.remarks) == 0)
	{
		ret = true;
	}
	return ret;
}

//operator!=
bool Account::operator!=(const Account& other)
{
	bool ret = false;
	if (this->date != other.date || this->briefs.compare(other.briefs) != 0 ||
		this->amount != other.amount || this->balance != other.balance ||
		this->remarks.compare(other.remarks) != 0)
	{
		ret = true;
	}
	return ret;
}

File: AccountBook/Java-AccountBook/src/AccountBook/Account.java
package AccountBook;

import java.text.NumberFormat;
import java.util.Currency;
import java.util.Locale;

public abstract class Account
{
    //인스턴스 필드 멤버
    protected Date date;
    protected String briefs;
    protected int amount;
    protected int balance;
    protected String remarks;

    //디폴트 생성자
    public Account()
    {
        this.date = new Date();
        this.briefs = "";
        this.amount = 0;
        this.balance = 0;
        this.remarks = "";
    }

    //매개변수를 가지는 생성자
    public Account(Date date, String briefs, int amount, int balance, String remarks)
    {
        this.date = date;
        this.briefs = briefs;
        this.amount = amount;
        this.balance = balance;
        this.remarks = remarks;
    }

    //equals 오버라이딩
    @Override
    public boolean equals(Object obj)
    {
        boolean ret = false;
        if(obj instanceof Account)
        {
            if(this.date.equals(((Account)obj).date) == true &&
                    this.briefs.compareTo(((Account)obj).briefs) == 0
                    && this.amount == ((Account)obj).amount &&
                    this.balance == ((Account)obj).balance &&
                    this.remarks.compareTo(((Account)obj).remarks) == 0)
            {
                ret = true;
            }
        }
        return ret;
    }

    //toString 오버라이딩
    @Override
    public String toString()
    {
        //통화에 원화표시를 해주고, 세자리수마다 콤마(,)를 찍어줌.
        String amount = NumberFormat.getCurrencyInstance(Locale.KOREA).format(this.amount);
        String balance = NumberFormat.getCurrencyInstance(Locale.KOREA).format(this.balance);
        return new String("일자 : " + this.date.toString() + ", 적요 : "
                + this.briefs + ", 금액 : " + amount + ", 잔액 : "
                + balance + ", 비고 : " + this.remarks);
    }

    //추상메소드 clone(자손인 Income이나 Outgo에서 구현함)
    public abstract Account clone();

    //잔액(balance) 계산하기 메소드
    public abstract int calculateBalance(Account account);
    public abstract int calculateBalance(int amount);

    //getter
    public Date getDate() {return this.date;}
    public String getBriefs() {return this.briefs;}
    public int getAmount() {return this.amount;}
    public int getBalance() {return this.balance;}
    public String getRemarks() {return this.remarks;}

    //setter
    //추상메소드 setAmount(자손인 Income이나 Outgo에서 구현함)
    public abstract void setAmount(int amount);
    public void setBalance(int balance)
    {
        this.balance = balance;
    }
}
